### PR TITLE
[#11572] Notification Feature - E2E Tests for User Notifications Page

### PR DESCRIPTION
**Pull request by zhaojj2209**
Part of #11572 

Adds E2E tests for instructor and student notifications page.

### CODE REVIEW

File: src/e2e/resources/data/InstructorNotificationsPageE2ETest.json [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      }
+    }
+  },
+  "notifications": {
+    "notification1": {
+      "notificationId": "notification1",
+      "startTime": "2011-01-01T00:00:00Z",
+      "endTime": "2099-01-01T00:00:00Z",
+      "createdAt": "2011-01-01T00:00:00Z",
+      "updatedAt": "2011-01-01T00:00:00Z",
```
Comment (ziqing26) [inline src/e2e/resources/data/InstructorNotificationsPageE2ETest.json:52]: `updatedAt` might not be necessary for test data now, considering it is transient in NotificationAttribute now. This also applies to the typical data bundle.

File: src/e2e/java/teammates/e2e/cases/InstructorNotificationsPageE2ETest.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @Test
+    @Override
+    public void testAll() {
+        AppUrl notificationsPageUrl = createFrontendUrl(Const.WebPageURIs.INSTRUCTOR_NOTIFICATIONS_PAGE);
+        InstructorNotificationsPage notificationsPage = loginToPage(notificationsPageUrl, InstructorNotificationsPage.class,
+                account.getGoogleId());
+
+        ______TS("verify that only the correct notifications are shown");
```
Comment (ziqing26) [inline src/e2e/java/teammates/e2e/cases/InstructorNotificationsPageE2ETest.java:38]: Would it be clearer if we add a bit more description here? For example:
```suggestion
        ______TS("verify that only active notifications with correct target user are shown");
```

**Review by ziqing26 — APPROVED** [review]
Overall LGTM!

File: src/e2e/java/teammates/e2e/cases/StudentNotificationsPageE2ETest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import teammates.common.util.AppUrl;
+import teammates.common.util.Const;
+import teammates.e2e.pageobjects.StudentNotificationsPage;
+
+/**
+ * SUT: {@link Const.WebPageURIs#STUDENT_NOTIFICATIONS_PAGE}.
+ */
+public class StudentNotificationsPageE2ETest extends BaseE2ETestCase {
+    private AccountAttributes account;
+    private NotificationAttributes notificationToMarkAsRead;
```
Comment (jianhandev) [inline src/e2e/java/teammates/e2e/cases/StudentNotificationsPageE2ETest.java:20]: These can be local variables

File: src/e2e/java/teammates/e2e/cases/InstructorNotificationsPageE2ETest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import teammates.common.util.AppUrl;
+import teammates.common.util.Const;
+import teammates.e2e.pageobjects.InstructorNotificationsPage;
+
+/**
+ * SUT: {@link Const.WebPageURIs#INSTRUCTOR_NOTIFICATIONS_PAGE}.
+ */
+public class InstructorNotificationsPageE2ETest extends BaseE2ETestCase {
+    private AccountAttributes account;
+    private NotificationAttributes notificationToMarkAsRead;
```
Comment (jianhandev) [inline src/e2e/java/teammates/e2e/cases/InstructorNotificationsPageE2ETest.java:20]: These can be local variables

**Review by jianhandev — APPROVED** [review]
LGTM

File: src/web/app/components/user-notifications-list/user-notifications-list.component.html [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<div class="row mb-3" *ngIf="notificationTabs.length > 0">
+<div class="row mb-3" [id]="timezone" *ngIf="notificationTabs.length > 0">
```
Comment (wkurniawan07) [inline src/web/app/components/user-notifications-list/user-notifications-list.component.html:1]: Any reason why the timezone is used as ID?

File: src/e2e/resources/data/InstructorNotificationsPageE2ETest.json [hunk 6]
```diff
@@ -0,0 +1,103 @@
+{
+  "accounts": {
+    "INotifications.instr": {
+      "googleId": "tm.e2e.INotifications.instr",
```
Comment (wkurniawan07) [inline src/e2e/resources/data/InstructorNotificationsPageE2ETest.json:4]: `Notifications` is a bit too long. Let's add a naming rule to shorten this to `Notifs`.

File: src/e2e/java/teammates/e2e/pageobjects/StudentNotificationsPage.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+import teammates.common.datatransfer.NotificationStyle;
+import teammates.common.datatransfer.attributes.NotificationAttributes;
+
+/**
+ * Page Object Model for student notifications page.
+ */
+public class StudentNotificationsPage extends AppPage {
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/StudentNotificationsPage.java:22]: As far as I can tell, this and `InstructorNotificationsPage` are duplicates of each other. There should be one base page object where both inherits from.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/e2e/java/teammates/e2e/pageobjects/StudentNotificationsPage.java (hunk 7 above)
Comment (zhaojj2209) [reply src/e2e/java/teammates/e2e/pageobjects/StudentNotificationsPage.java:22]: Got it, will create another page object `UserNotificationsPage` and have `InstructorNotificationsPage` and `StudentNotificationsPage` inherit from it

File: src/web/app/components/user-notifications-list/user-notifications-list.component.html (hunk 5 above)
Comment (zhaojj2209) [reply src/web/app/components/user-notifications-list/user-notifications-list.component.html:1]: This is so that the E2E tests can access the timezone that is used to format the start/end date for the notification. I've put it as an ID here so that it won't be shown to the user; should I add a line to display the timezone to the user instead (e.g. `All dates are displayed in {timezone} time.`), similar to the admin notifications page?

File: src/e2e/java/teammates/e2e/pageobjects/UserNotificationsPage.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        assertEquals(getHeaderText(notification), cardHeader.getText());
+        assertTrue(cardHeader.getAttribute("class").contains(getHeaderClass(notification.getStyle())));
+
+        // Checks if tab is open if notification is unread, and closed if notification is read
+        String chevronClass = notificationTab.findElement(By.tagName("i")).getAttribute("class");
+        if (isRead) {
+            assertTrue(chevronClass.contains("fa-chevron-down"));
+            // Open tab if notification is unread
+            click(cardHeader);
+            waitForPageToLoad();
```
Comment (NicolasCwy) [inline src/e2e/java/teammates/e2e/pageobjects/UserNotificationsPage.java:70]: I believe `waitForPageLoad()` waits for the animation to finish, so `waitUntilAnimationFinish` can be removed

File: src/e2e/java/teammates/e2e/pageobjects/UserNotificationsPage.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        List<WebElement> markAsReadBtnList = notificationTab.findElements(By.id("btn-mark-as-read"));
+
+        if (isRead) {
+            // Check that mark as read button cannot be found if notification is read
+            assertEquals(0, markAsReadBtnList.size());
+
+            // Close tab if notification is read
+            click(cardHeader);
+            waitForPageToLoad();
+            waitUntilAnimationFinish();
```
Comment (NicolasCwy) [inline src/e2e/java/teammates/e2e/pageobjects/UserNotificationsPage.java:89]: Same here

**Review by NicolasCwy — APPROVED** [review]
LGTM!

**Review by fsgmhoward — APPROVED** [review]

File: src/e2e/java/teammates/e2e/pageobjects/InstructorNotificationsPage.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+ */
+public class InstructorNotificationsPage extends UserNotificationsPage {
+
+    public InstructorNotificationsPage(Browser browser) {
+        super(browser);
+    }
+
+    @Override
+    protected boolean containsExpectedPageContents() {
+        return getPageSource().contains("Notifications");
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/InstructorNotificationsPage.java:14]: Should this be "Instructor Notifications"?

**Review by wkurniawan07 — APPROVED** [review]
LGTM with one minor comment

File: src/e2e/java/teammates/e2e/pageobjects/InstructorNotificationsPage.java (hunk 10 above)
Comment (zhaojj2209) [reply src/e2e/java/teammates/e2e/pageobjects/InstructorNotificationsPage.java]: Thanks for the catch! I've pushed a quick fix
